### PR TITLE
Directed energy deposition examples

### PR DESCRIPTION
**Pull request by dewenyushu**
- Add examples for simulating directed energy deposition process
- Add velocity and function path Gaussian heat source terms
- Add associated tests and documentation

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -36,7 +36,7 @@ FUNCTIONAL_EXPANSION_TOOLS  := no
 GEOCHEMISTRY                := no
 HEAT_CONDUCTION             := yes
 LEVEL_SET                   := yes
-MISC                        := no
+MISC                        := yes
```
Comment (jiangwen84) [inline Makefile:39]: why do you need this module?

File: examples/directed_energy_deposition/cylinder/master_app_mechanical.i [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+    block = 2
+    new_boundary = 'moving_boundary'
+  []
+
+  [cmbn]
+    type = CombinerGenerator
+    inputs = 'add_set2 moving_boundary'
+  []
+
+  displacements = 'disp_x disp_y disp_z'
```
Comment (jiangwen84) [inline examples/directed_energy_deposition/cylinder/master_app_mechanical.i]: remove it as it is already defined in GlobalParam?

File: examples/directed_energy_deposition/cylinder/master_app_mechanical.i [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+  line_search = 'none'
+
+  l_max_its = 100
+  nl_max_its = 20
+  nl_rel_tol = 1e-6
+  nl_abs_tol = 1e-8
+
+  start_time = 0.0
+  end_time = 57036
+  dt = ${dt} # ms
```
Comment (jiangwen84) [inline examples/directed_energy_deposition/cylinder/master_app_mechanical.i:294]: since you give the unit of time here, could you also add the units for length?

File: examples/directed_energy_deposition/cylinder/sub_app_thermal.i [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+    scale_factor = 1.0
+  []
+  [thermal_conductivity_metal]
+    type = PiecewiseLinear
+    x = '-1e7 198.84  298.10  398.75 500.76 601.40 700.64 801.27 901.89 1001.12 1098.98 1200.96 '
+        '1301.56 1400.78 1501.37 1601.96 1e7'
+    y = '247.72 247.72 285.64 323.55 358.44 390.29 417.59 446.41 469.16 491.91 510.11 528.31 540.44 '
+        '554.09 561.67 569.26 569.26'
+    format = columns
+    scale_factor = 0.05e-6
```
Comment (jiangwen84) [inline examples/directed_energy_deposition/cylinder/sub_app_thermal.i:198]: why did you add those scale_factor?

File: examples/directed_energy_deposition/cylinder/master_app_mechanical.i [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+[Preconditioning]
+  [smp]
+    type = SMP
+    full = true
+  []
+[]
+
+[Executioner]
+  type = Transient
+  solve_type = 'PJFNK'
```
Comment (jiangwen84) [inline examples/directed_energy_deposition/cylinder/master_app_mechanical.i]: NEWTON?

File: examples/directed_energy_deposition/cylinder/sub_app_thermal.i [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+[Preconditioning]
+  [smp]
+    type = SMP
+    full = true
+  []
+[]
+
+[Executioner]
+  type = Transient
+  solve_type = 'PJFNK'
```
Comment (jiangwen84) [inline examples/directed_energy_deposition/cylinder/sub_app_thermal.i]: NEWTON?

File: examples/directed_energy_deposition/hollow_cylinder/master_app_mechanical.i [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+  #   new_boundary = 'middle'
+  #   normal = '0 0 1'
+  # []
+
+  [cmbn]
+    type = CombinerGenerator
+    inputs = 'add_set2 moving_boundary'
+  []
+
+  displacements = 'disp_x disp_y disp_z'
```
Comment (jiangwen84) [inline examples/directed_energy_deposition/hollow_cylinder/master_app_mechanical.i]: same here

File: examples/directed_energy_deposition/hollow_cylinder/master_app_mechanical.i [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+    type = ADDiffusion
+    variable = 'disp_z'
+    block = 1
+  []
+[]
+
+[Modules/TensorMechanics/Master]
+  strain = FINITE
+  incremental = true
+  add_variables = true
```
Comment (jiangwen84) [inline examples/directed_energy_deposition/hollow_cylinder/master_app_mechanical.i]: add_variable = true, but you still manually add those variables. Keep one.

File: examples/directed_energy_deposition/hollow_cylinder/sub_app_thermal.i [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+    block = 2
+    new_boundary = 'moving_boundary'
+  []
+  # [middle]
+  #   type = SideSetsAroundSubdomainGenerator
+  #   input = moving_boundary
+  #   block = 3
+  #   new_boundary = 'middle'
+  #   normal = '0 0 1'
+  # []
```
Comment (jiangwen84) [inline examples/directed_energy_deposition/hollow_cylinder/sub_app_thermal.i:58]: do we need to keep those?

File: include/materials/FunctionPathGaussianHeatSource.h [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+/*                           ALL RIGHTS RESERVED                            */
+/****************************************************************************/
+
+#pragma once
+
+#include "GaussianHeatSourceBase.h"
+#include "Function.h"
+
+/**
+ * Double ellipsoid heat source distribution.
```
Comment (jiangwen84) [inline include/materials/FunctionPathGaussianHeatSource.h]: Gaussian or double ellipsoid?

File: src/materials/VelocityGaussianHeatSource.C [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+  params.addParam<FunctionName>(
+      "function_vy",
+      0,
+      "The function of y component of the center of the heating spot moving speed");
+  params.addParam<FunctionName>(
+      "function_vz",
+      0,
+      "The function of z component of the center of the heating spot moving speed");
+
+  params.addClassDescription("Double ellipsoid volumetric source heat with moving velocity.");
```
Comment (jiangwen84) [inline src/materials/VelocityGaussianHeatSource.C]: Is the description accurate?

File: src/materials/VelocityGaussianHeatSource.C [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+VelocityGaussianHeatSource::validParams()
+{
+  InputParameters params = GaussianHeatSourceBase::validParams();
+  params.addParam<Real>("x0", 0, "The x component of the initial center of the heating spot");
+  params.addParam<Real>("y0", 0, "The y component of the initial center of the heating spot");
+  params.addParam<Real>("z0", 0, "The z component of the initial center of the heating spot");
+
+  params.addParam<FunctionName>(
+      "function_vx",
+      0,
```
Comment (jiangwen84) [inline src/materials/VelocityGaussianHeatSource.C]: should be the default function name "0"?

File: src/materials/VelocityGaussianHeatSource.C [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+{
+}
+
+void
+VelocityGaussianHeatSource::computeHeatSourceCenterAtTime(Real & x,
+                                                          Real & y,
+                                                          Real & z,
+                                                          const Real & time)
+{
+  Real delta_t;
```
Comment (jiangwen84) [inline src/materials/VelocityGaussianHeatSource.C]: you can move delta_t into "if" statement.

File: src/materials/GaussianHeatSourceBase.C [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+                        "option to use user input effective radii or from experimentally fitted "
+                        "formulations. Default is to use user input data.");
+  params.addParam<std::vector<Real>>(
+      "r",
+      "effective radii along three directions. If only one parameter is provided, then we assume "
+      "the effective radius to be equal along three directions.");
+
+  params.addParam<Real>(
+      "feed_rate",
+      0.000124,
```
Comment (jiangwen84) [inline src/materials/GaussianHeatSourceBase.C]: 0.000124 seems to be a bit arbitrary. It may be better to put 0.0 here.

File: src/materials/GaussianHeatSourceBase.C [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+    _P(getParam<Real>("power")),
+    _eta(getParam<Real>("efficiency")),
+    _feed_rate(getParam<Real>("feed_rate")),
+    _r(getParam<std::vector<Real>>("r")),
+    _f(getParam<Real>("factor")),
+    _std_factor(getParam<Real>("std_factor")),
+    _heat_source_type(getParam<MooseEnum>("heat_source_type").getEnum<HeatSourceType>()),
+    _threshold_length(getParam<Real>("threshold_length")),
+    _number_time_integration(getParam<Real>("number_time_integration")),
+    _volumetric_heat(declareADProperty<Real>("volumetric_heat"))
```
Comment (jiangwen84) [inline src/materials/GaussianHeatSourceBase.C]: If this is the AD version, should we rename the class to ADxxx?

File: src/materials/GaussianHeatSourceBase.C [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+         std::exp(dist_x + dist_y + dist_z);
+}
+
+Real
+GaussianHeatSourceBase::computeAveragedHeatSource(
+    const Real x, const Real y, const Real z, const Real time_begin, const Real time_end)
+{
+  mooseAssert(time_end > time_begin, "Begin time should be smaller than end time.");
+
+  unsigned int num_pts = 5;
```
Comment (jiangwen84) [inline src/materials/GaussianHeatSourceBase.C]: why 5?

File: src/materials/GaussianHeatSourceBase.C [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+  // we do not update _r if we do not proceed in time
+  if (time <= _r_time_prev)
+    return;
+  else
+    _r_time_prev = time;
+
+  // get scaled laser power, scanning speed, and powder feed rate
+  Real lp = _P / 1.0e-3 / 400.0;             // input in unit 1e-3 W
+  Real ss = _scan_speed / 4.23333e-4 / 40.0; // input in mm/ms (1 ipm = 4.23333e-4 mm/ms)
+  Real pf = _feed_rate / 0.031e-3 / 15.0;    // input in g/ms (1rpm = 0.031e-3 g/ms)
```
Comment (jiangwen84) [inline src/materials/GaussianHeatSourceBase.C]: looks like you used a specific unit system. Did you add documents for those and description in the InputParams?

File: src/materials/GaussianHeatSourceBase.C [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+    // compute the material bead height (z dimension)
+    mean_rz += _height_param[i] * vals[i];
+  }
+
+  // sample the r[0] and r[2] value
+  // define a random number generator
+  std::random_device rd{};
+  std::mt19937 generator{rd()};
+  std::normal_distribution<double> dist_xy(mean_rxy, std_r);
+  std::normal_distribution<double> dist_z(mean_rz, std_r);
```
Comment (jiangwen84) [inline src/materials/GaussianHeatSourceBase.C]: Is it a general capability that you would like to add for this object? Or is it specific to your LDRD project? Should we add an option to use the deterministic value?

File: include/materials/GaussianHeatSourceBase.h [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+  const std::vector<Real> _height_param = {-0.01697178,
+                                           -0.61333238,
+                                           0.31978383,
+                                           0.15406738,
+                                           0.891539,
+                                           -0.32619461,
+                                           -0.78659714,
+                                           1.18223877,
+                                           -0.55376121,
+                                           0.19044780180084175};
```
Comment (jiangwen84) [inline include/materials/GaussianHeatSourceBase.h]: What are those numbers from? Should we hard-code them in the code?

**Review by jiangwen84 — CHANGES_REQUESTED** [review]
Looks good. A few minor comments.

File: Makefile (hunk 1 above)
Comment (dewenyushu) [reply Makefile:39]: To use `ADDensity`

File: examples/directed_energy_deposition/cylinder/sub_app_thermal.i (hunk 4 above)
Comment (dewenyushu) [reply examples/directed_energy_deposition/cylinder/sub_app_thermal.i:198]: This is to match the value from the literature. I'll add a comment

File: examples/directed_energy_deposition/hollow_cylinder/sub_app_thermal.i (hunk 9 above)
Comment (dewenyushu) [reply examples/directed_energy_deposition/hollow_cylinder/sub_app_thermal.i:58]: Good catch, I'll remove them

File: src/materials/VelocityGaussianHeatSource.C (hunk 12 above)
Comment (dewenyushu) [reply src/materials/VelocityGaussianHeatSource.C]: I'll change the default to 'vx', 'vy' and 'vz'.

File: src/materials/VelocityGaussianHeatSource.C (hunk 11 above)
Comment (dewenyushu) [reply src/materials/VelocityGaussianHeatSource.C]: I'll update that

File: src/materials/VelocityGaussianHeatSource.C (hunk 12 above)
Comment (dewenyushu) [reply src/materials/VelocityGaussianHeatSource.C]: ^ no to above. Both "0" and 0 should work. I'll change to "0"

**Review by cticenhour — DISMISSED** [review]
One general comment, since I think @jiangwen84 has the content of the rest of the PR handled. If placing inputs in `examples`, I would like to start adding syntax check tests to those inputs so they don't get broken over time. Thank you for this contribution!

File: src/materials/GaussianHeatSourceBase.C (hunk 16 above)
Comment (dewenyushu) [reply src/materials/GaussianHeatSourceBase.C]: It's the number of how many time steps to use for integration. It will double until the integral reaches good accuracy or reaches the limit of time step counts. I'll add to comment about this.

File: src/materials/GaussianHeatSourceBase.C (hunk 18 above)
Comment (dewenyushu) [reply src/materials/GaussianHeatSourceBase.C]: Yes, by default the deterministic values will be utilized. It will be used only when `use_input_r = false` and `std_factor` !=0

File: include/materials/GaussianHeatSourceBase.h (hunk 19 above)
Comment (dewenyushu) [reply include/materials/GaussianHeatSourceBase.h]: These are coefficients fitted from experimental measurements. This is to include feed-rate impact. I mentioned this in the documentation. I'll add a comment here, too.

File: src/materials/GaussianHeatSourceBase.C (hunk 17 above)
Comment (dewenyushu) [reply src/materials/GaussianHeatSourceBase.C]: Will do

**Comment by dewenyushu** [comment]
> One general comment, since I think @jiangwen84 has the content of the rest of the PR handled. If placing inputs in `examples`, I would like to start adding syntax check tests to those inputs so they don't get broken over time. Thank you for this contribution!

Thanks for the gentle reminder! The syntax checks for the examples are up!

**Review by jiangwen84 — APPROVED** [review]
looks great. Thanks for your contribution.
